README.md
# ALGO PARA EDITAR JSON

La verdad no sé que es esto, te edita un archivo donde puedes guardar usuarios y tal --pero solo puedes guardar el nombre y la edad-- Y PUEDES GUARDAR LO Q QUIERAS. Las ids se cambian solas y no sé, es algo para practicar


docs/changelog.md
# Changelog
 
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/),
and this project adheres to [Semantic Versioning](https://semver.org/).

## [Unreleased]

The objective of this section is to show the upcoming modifications.


## [1.1.1] - 2023-03-05

### Added

- Arabic translation (#444).
- v1.1 French translation.
- v1.1 Dutch translation (#371).
- v1.1 Russian translation (#410).
- v1.1 Japanese translation (#363).
- v1.1 Norwegian Bokmål translation (#383).
- v1.1 "Inconsistent Changes" Turkish translation (#347).
- Default to most recent versions available for each languages
- Display count of available translations (26 to date!)
- Centralize all links into `/data/links.json` so they can be updated easily

### Fixed

- Improve French translation (#377).
- Improve id-ID translation (#416).
- Improve Persian translation (#457).
- Improve Russian translation (#408).
- Improve Swedish title (#419).
- Improve zh-CN translation (#359).
- Improve French translation (#357).
- Improve zh-TW translation (#360, #355).
- Improve Spanish (es-ES) transltion (#362).
- Foldout menu in Dutch translation (#371).
- Missing periods at the end of each change (#451).
- Fix missing logo in 1.1 pages
- Display notice when translation isn't for most recent version
- Various broken links, page versions, and indentations.

### Changed

- Upgrade dependencies: Ruby 3.2.1, Middleman, etc.

### Removed

- Unused normalize.css file
- Identical links assigned in each translation file
- Duplicate index file for the english version

## [1.1.0] - 2019-02-15

### Added

- Danish translation (#297).
- Georgian translation from (#337).
- Changelog inconsistency section in Bad Practices.

### Fixed

- Italian translation (#332).
- Indonesian translation (#336).

## [1.0.0] - 2017-06-20

### Added

- New visual identity by [@tylerfortune8](https://github.com/tylerfortune8).
- Version navigation.
- Links to latest released version in previous versions.
- "Why keep a changelog?" section.
- "Who needs a changelog?" section.
- "How do I make a changelog?" section.
- "Frequently Asked Questions" section.
- New "Guiding Principles" sub-section to "How do I make a changelog?".
- Simplified and Traditional Chinese translations from [@tianshuo](https://github.com/tianshuo).
- German translation from [@mpbzh](https://github.com/mpbzh) & [@Art4](https://github.com/Art4).
- Italian translation from [@azkidenz](https://github.com/azkidenz).
- Swedish translation from [@magol](https://github.com/magol).
- Turkish translation from [@emreerkan](https://github.com/emreerkan).
- French translation from [@zapashcanon](https://github.com/zapashcanon).
- Brazilian Portuguese translation from [@Webysther](https://github.com/Webysther).
- Polish translation from [@amielucha](https://github.com/amielucha) & [@m-aciek](https://github.com/m-aciek).
- Russian translation from [@aishek](https://github.com/aishek).
- Czech translation from [@h4vry](https://github.com/h4vry).
- Slovak translation from [@jkostolansky](https://github.com/jkostolansky).
- Korean translation from [@pierceh89](https://github.com/pierceh89).
- Croatian translation from [@porx](https://github.com/porx).
- Persian translation from [@Hameds](https://github.com/Hameds).
- Ukrainian translation from [@osadchyi-s](https://github.com/osadchyi-s).

### Changed

- Start using "changelog" over "change log" since it's the common usage.
- Start versioning based on the current English version at 0.3.0 to help
  translation authors keep things up-to-date.
- Rewrite "What makes unicorns cry?" section.
- Rewrite "Ignoring Deprecations" sub-section to clarify the ideal
  scenario.
- Improve "Commit log diffs" sub-section to further argument against
  them.
- Merge "Why can’t people just use a git log diff?" with "Commit log
  diffs".
- Fix typos in Simplified Chinese and Traditional Chinese translations.
- Fix typos in Brazilian Portuguese translation.
- Fix typos in Turkish translation.
- Fix typos in Czech translation.
- Fix typos in Swedish translation.
- Improve phrasing in French translation.
- Fix phrasing and spelling in German translation.

### Removed

- Section about "changelog" vs "CHANGELOG".

src/main.py
import os

import funciones

if __name__ != "__main__":
    print("Que haces tonto")
    quit()

rutaArchivoJSON = os.path.dirname(__file__) + "/004 - archivo.json"

def reload():
    contenido = funciones.cargarJSON(rutaArchivoJSON)
    usuarios = contenido["main"]
    usuariosCopia = contenido["main"].copy()
    usuariosEliminar = []
    config = contenido["config"]
    parametros = config["parametros"]
    return contenido, usuarios, usuariosCopia, usuariosEliminar, config, parametros

contenido, usuarios, usuariosCopia, usuariosEliminar, config, parametros = reload()

posiblesRespuestas = ["l", "list", "n", "new", "r", "remove", "s", "save", "q", "quit", "cp", "cparameter", "createparameter"]
eleccion = funciones.verComandos()

if(eleccion not in posiblesRespuestas):
    eleccion = input("ERROR! Deberías escribir algo valido: ")

while True:
    
    if(eleccion == "cp" or eleccion == "cparameter" or eleccion == "createparameter"):
        nombreParametro = input("Introduce el nombre del parámetro que deseas crear.\n > ").lower()
        while any([i[1] == nombreParametro for i in parametros]):
            nombreParametro = input("ERROR! Ya existe un parámetro con ese nombre. Introduce otro.\n > ").lower()

        inputParametro = input("Introduce el mensaje para pedir datos del parámetro que deseas crear.\nP.ej: Introduce la ciudad de nacimiento.\n > ")
        
        tiposPermitidos = ["str", "string", "int", "integer", "numero", "float", "decimal", "array", "lista", "boolean", "bool", "dict", "dictionary", "diccionario", "cualquiera", "any", ""]
        tipoParametro = input("Introduce el tipo de valor permitido para el parámetro.\n > ").lower()
        while not tipoParametro in tiposPermitidos:
            tipoParametro = input("ERROR! Ese tipo no está permitido. Selecciona uno de la siguiente lista:\nstr (string), int (numero), float (decimal), array (lista), bool (booleano), dict (diccionario) o any (cualquiera).\n > ").lower()
        
        if tipoParametro in ["str", "string"]:
            tipoParametro = "str"
        elif tipoParametro in ["int", "integer", "numero"]:
            tipoParametro = "int"
        elif tipoParametro in ["float", "decimal",]:
            tipoParametro = "float"
        elif tipoParametro in ["array", "lista"]:
            tipoParametro = "list"
        elif tipoParametro in ["bool", "boolean"]:
            tipoParametro = "bool"
        elif tipoParametro in ["dict", "dictionary", "diccionario"]:
            tipoParametro = "dict"
        else: # any
            tipoParametro = "None"
        
        respuestasPermitidas = ["sí", "si", "s", "yes", "y", "", "no", "n"]
        parametroObligatorio = input("Es obligatorio introducir este parámetro? (S/n)\n > ").lower()
        while not parametroObligatorio in respuestasPermitidas:
            parametroObligatorio = input("ERROR! Debes introducir si o no.\n > ").lower()
        
        if parametroObligatorio in ["sí", "si", "s", "yes", "y", ""]:
            parametroObligatorio = True
        else:
            parametroObligatorio = False

        funciones.crearParametro(parametros, nombreParametro, inputParametro, parametroObligatorio, tipoParametro)

    if(eleccion == "l" or eleccion == "list"):
        funciones.verUsuarios(usuarios, usuariosEliminar, usuariosCopia)

    if(eleccion == "n" or eleccion == "new"):
        listaParametros = []
        for i in parametros:
            valor = input(i[2] + (' (Opcional)' if not eval(i[0]) else '') + "\n > ")
            try:
                if valor != "":
                    tipoValor = valor.split()[0] # Para que no pete si metes dos palabras separadas
                    tipoValor = type(eval(tipoValor))
                else:
                    tipoValor = str
            except NameError:
                tipoValor = str

            if bool(i[0]):
                if not i[3] == "None":
                    while tipoValor != eval(i[3]):
                        valor = input(f"ERROR! El valor introducido es de tipo {tipoValor} pero requiere un {i[3]}.\n > ")
            listaParametros.append([i[1], valor])

        funciones.crearUsuario(contenido, listaParametros)

    if(eleccion == "r" or eleccion == "remove"):
        identificador = input("Introduce un ID\n > ")
        while not identificador.isdigit() or not int(identificador) in funciones.ids:
            while not identificador.isdigit():
                identificador = input("Debes introducir un identificador.\n > ")
            while not int(identificador) in funciones.ids:
                identificador = input("ERROR! Esa ID no existe. Introduce una existente.\n > ")
        usuariosEliminar.append(int(identificador))

    if(eleccion == "s" or eleccion == "save"):
        funciones.guardarJSON(rutaArchivoJSON, contenido, usuariosEliminar)
        contenido, usuarios, usuariosCopia, usuariosEliminar, config, parametros = reload()

    if(eleccion == "q" or eleccion == "quit"):
        print("deu")
        quit()

    if(eleccion not in posiblesRespuestas):
        eleccion = input("ERROR! Deberías escribir algo valido: ")
    else:
        eleccion = funciones.verComandos()


src/funciones.py
import json

ids = []
fileName = "file" # Quiero hacer que puedas cambiar el nombre del archivo json

def tutorial():
    print("Bienvenido a lo que sea esto (tener en cuenta que esto solo aparece cuando el archivo json no existe)")
    print("Se recomienda configurar primero los parámetros y luego no tocarlos para que no hayan errores")
    print("Se explica algo mas y ya")

def cargarJSON(rutaArchivo: str):
    try: # Si encuentra el archivo
        with open(rutaArchivo, "r") as archivo:
            contenido = json.loads(archivo.read())
    except FileNotFoundError: # Si no encuentra el archivo crearlo
        with open(rutaArchivo, "w") as archivo:
            tutorial()
            archivo.write('{"main":[],"config":{"parametros":[]}}')
            contenido = json.loads('{"main":[],"config":{"parametros":[]}}')
    except json.decoder.JSONDecodeError: # Si el archivo está vacio editarlo
        with open(rutaArchivo, "w") as archivo:
            archivo.write('{"main":[],"config":{"parametros":[]}}')
            contenido = json.loads('{"main":[],"config":{"parametros":[]}}')
    for usuario in contenido["main"]:
        ids.append(usuario["id"])
    return contenido

def guardarJSON(rutaArchivo: str, contenido: str, usuariosEliminar: list):
    for id in usuariosEliminar:
        eliminarUsuario(contenido["main"], id)
    with open(rutaArchivo, "w") as archivo:
        archivo.write(json.dumps(contenido))
    

def crearUsuario(contenido: list, parametros: list):
    id = 1
    while id in ids:
        id += 1
    ids.append(id)
    usuario = {"id": id}
    for parametro in parametros:
        usuario[parametro[0]] = parametro[1]
    contenido["main"].append(usuario)


def eliminarUsuario(listaUsuarios: list, id: int):
    result = next((usuario for usuario in listaUsuarios if usuario["id"] == id), None)
    if result:
        listaUsuarios.remove(result)
        ids.remove(id)


def verUsuarios(usuarios: list, usuariosEliminar: list, usuariosCopia: list):
    for usuario in usuarios:
        if usuario["id"] in usuariosEliminar:
            print("\33[41m" + str(usuario) + "\33[0m")
        elif usuario not in usuariosCopia:
            print("\33[42m" + str(usuario) + "\33[0m")
        else:
            print(usuario)

def verComandos():
    print("Puedes usar varios comandos")
    print("cp - Para crear un nuevo parámetro")
    print("rp - Para eliminar un parámetro")
    print()
    print("l - Para ver una lista de usuarios")
    print("n - Para crear un nuevo usuario")
    print("r - Para eliminar un usuario existente")
    print("s - Para guardar la lista actual de usuarios")
    print("q - Para abandonar esta terminal")
    return input(" > ")

def crearParametro(listaParametros: list, parametro: str, pregunta:str, obligatorio: bool, tipo: type):
    listaParametros.append([str(obligatorio), parametro, pregunta, str(tipo)])


004 - ModificarJSON.py
import os
import json

def cargarJSON(rutaArchivo):
    global usuariosCopia
    try:
        with open(rutaArchivo, "r") as archivo:
            contenido = json.loads(archivo.read())
    except FileNotFoundError:
        with open(rutaArchivo, "w") as archivo:
            archivo.write('{"usuarios":[],"bots":[]}')
            contenido = json.loads('{"usuarios":[],"bots":[]}')
    for usuario in contenido["usuarios"]:
        ids.append(usuario["id"])
    usuariosCopia = contenido["usuarios"].copy()
    return contenido

def guardarJSON(rutaArchivo):
    global usuariosCopia
    global usuariosEliminar
    for id in usuariosEliminar:
        eliminarUsuario(id)
    with open(rutaArchivo, "w") as archivo:
        archivo.write(json.dumps(contenido))
    usuariosCopia = contenido["usuarios"].copy()
    usuariosEliminar = []
    

def crearUsuario(nombre: str, edad: int):
    id = 1
    while id in ids:
        id += 1
    ids.append(id)
    usuarios.append({'id': id, 'nombre': nombre, 'edad': edad})


def eliminarUsuario(id: int):
    result = next((usuario for usuario in usuarios if usuario["id"] == id), None)
    if result:
        usuarios.remove(result)
        ids.remove(id)


def verUsuarios():
    for usuario in usuarios:
        if usuario["id"] in usuariosEliminar:
            print("\33[41m" + str(usuario) + "\33[0m")
        elif usuario not in usuariosCopia:
            print("\33[42m" + str(usuario) + "\33[0m")
        else:
            print(usuario)

def verComandos():
    print("Puedes usar varios comandos")
    print("l - Para ver una lista de usuarios")
    print("n - Para crear un nuevo usuario")
    print("r - Para eliminar un usuario existente")
    print("s - Para guardar la lista actual de usuarios")
    print("q - Para abandonar esta terminal")

rutaArchivoJSON = os.path.dirname(__file__) + "\\004 - archivo.json"
ids = []

contenido = cargarJSON(rutaArchivoJSON)
usuarios = contenido["usuarios"]
usuariosCopia = contenido["usuarios"].copy()
usuariosEliminar = []

posiblesRespuestas = ["l", "list", "n", "new", "r", "remove", "s", "save", "q", "quit"]
verComandos()
eleccion = input(" > ")
if(eleccion not in posiblesRespuestas):
    eleccion = input("Deberías escribir algo valido: ")
while True:
    if(eleccion == "l" or eleccion == "list"):
        verUsuarios()

    if(eleccion == "n" or eleccion == "new"):
        nombre = input("Como quieres llamar a este usuario?\n > ")
        edad = input("Cual será la edad de este usuario?\n > ")
        while not edad.isdigit():
            edad = input(f"{edad} no es un nombre entero. Introduce un nombre entero.\n > ")
        crearUsuario(nombre, int(edad))

    if(eleccion == "r" or eleccion == "remove"):
        identificador = input("Introduce un ID\n > ")
        while not identificador.isdigit():
            identificador = input("Debes introducir un identificador.")
        usuariosEliminar.append(int(identificador))

    if(eleccion == "s" or eleccion == "save"):
        guardarJSON(rutaArchivoJSON)

    if(eleccion == "q" or eleccion == "quit"):
        print("deu")
        quit()

    if(eleccion not in posiblesRespuestas):
        eleccion = input("Deberías escribir algo valido: ")
    else:
        verComandos()
        eleccion = input(" > ")

